packages/profile-sync-controller/src/controllers/user-storage/__fixtures__/mockMessenger.ts

```diff
@@ -0,0 +1,287 @@
+import type {
+  AllowedActions,
+  AllowedEvents,
+  UserStorageControllerMessenger,
+} from '..';
+
+type GetHandler<ActionType extends AllowedActions['type']> = Extract<
+  AllowedActions,
+  { type: ActionType }
+>['handler'];
+
+type CallParams = {
+  [K in AllowedActions['type']]: [
+    K,
+    ...Parameters<Extract<AllowedActions, { type: K }>['handler']>,
+  ];
+}[AllowedActions['type']];
+
+const typedMockFn = <
+  ActionType extends AllowedActions['type'],
+  Handler = GetHandler<ActionType>,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Func extends (...args: any) => any = Handler extends (...args: any[]) => any
+    ? Handler
+    : never,
+>(
+  _type: ActionType,
+) => jest.fn<ReturnType<Func>, Parameters<Func>>();
+
+type ExternalEvents = NotNamespacedBy<
+  'UserStorageController',
+  AllowedEvents['type']
+>;
+
+/**
+ * creates a custom user storage messenger, in case tests need different permissions
+ * @param props - overrides
+ * @param props.overrideEvents - override events
+ * @returns base messenger, and messenger. You can pass this into the mocks below to mock messenger calls
+ */
+export function createCustomUserStorageMessenger(props?: {
+  overrideEvents?: ExternalEvents[];
+}) {
+  const baseMessenger = new ControllerMessenger<
+    AllowedActions,
+    AllowedEvents
+  >();
+  const messenger = baseMessenger.getRestricted({
+    name: 'UserStorageController',
+    allowedActions: [
+      'KeyringController:getState',
+      'KeyringController:addNewAccount',
+      'SnapController:handleRequest',
+      'AuthenticationController:getBearerToken',
+      'AuthenticationController:getSessionProfile',
+      'AuthenticationController:isSignedIn',
+      'AuthenticationController:performSignOut',
+      'AuthenticationController:performSignIn',
+      'NotificationServicesController:disableNotificationServices',
+      'NotificationServicesController:selectIsNotificationServicesEnabled',
+      'AccountsController:listAccounts',
+      'AccountsController:updateAccountMetadata',
+      'NetworkController:getState',
+      'NetworkController:addNetwork',
+      'NetworkController:updateNetwork',
+      'NetworkController:removeNetwork',
+    ],
+    allowedEvents: props?.overrideEvents ?? [
+      'KeyringController:lock',
+      'KeyringController:unlock',
+      'AccountsController:accountAdded',
+      'AccountsController:accountRenamed',
+      'NetworkController:networkRemoved',
+    ],
+  });
+
+  return {
+    baseMessenger,
+    messenger,
+  };
+}
+
+type OverrideMessengers = {
+  baseMessenger: ControllerMessenger<AllowedActions, AllowedEvents>;
+  messenger: UserStorageControllerMessenger;
+};
+
+/**
+ * Jest Mock Utility to generate a mock User Storage Messenger
+ * @param overrideMessengers - override messengers if need to modify the underlying permissions
+ * @returns series of mocks to actions that can be called
+ */
+export function mockUserStorageMessenger(
```

Comment:
reusable mock messenger.

The fact that this messenger has so many actions it requires is probably in indicator we need to split up the syncing logic into separate controllers.